File: src/qrs-core-js/src/services/verificationService.ts
/**
 * VerificationService: the verification pipeline.
 *
 * Pipeline (protocol §54 adapted):
 *   1. parse the SDoc
 *   2. resolve and verify the TCert (self-signature, id, validity)
 *   3. resolve trust (pinning or CA attestation)
 *   4. evaluate revocation (key, TCert, SDoc block)
 *   5. collect secret inputs, rebuild the COSE external AAD, verify the signature
 *   6. validate the payload against the schema (including contextual fields)
 *   7. produce a structured VerificationResult
 *
 * The result deliberately distinguishes `valid`, `invalid` and `cannotVerify`
 * (e.g. GPS unavailable is NOT the same as "outside the permitted area").
 */
import { cborEncode } from '../cbor/canonical.js';
import type { IContextProvider } from '../context/context.js';
import type { PublicJwk } from '../crypto/jwk.js';
import type { CryptoRegistry } from '../crypto/registry.js';
import { constantTimeEqual, toHex } from '../id.js';
import {
  effectiveBinding,
  isBoundField,
  isStrippedBinding,
  type FieldResult,
  type FieldSchema,
  type VerificationContext,
} from '../fields/types.js';
import type { FieldRegistry } from '../fields/registry.js';
import { QrsValidationError } from '../errors.js';
import { sdocIdOf, tcertIdOf, tcertNumberOf } from '../signedObject/signedObject.js';
import { parseSignedObject, verifyParsedSignedObject } from '../signedObject/signedObject.js';
import type { ICertificateStore, IEndpointConfigStore } from '../storage/stores.js';
import type { SdocId, TcertId } from '../types.js';
import type { IClock } from './clock.js';
import type { RevocationService } from './revocationService.js';
import type { ComponentState, TrustService } from './trustService.js';

export interface VerificationResult {
  overall: ComponentState;
  cryptographic: ComponentState;
  tcert: ComponentState;
  trust: ComponentState;
  revocation: ComponentState;
  schema: ComponentState;
  fields: FieldResult[];
  context: 'satisfied' | 'missing' | 'denied';
  warnings: string[];
  sdocId?: SdocId;
  tcertId?: TcertId;
  message?: string;
}

export interface VerifyOptions {
  /** Override "now" (epoch seconds) for deterministic verification. */
  currentTime?: number;
}

export interface VerificationServiceDeps {
  cryptoRegistry: CryptoRegistry;
  fieldRegistry: FieldRegistry;
  certificateStore: ICertificateStore;
  endpointConfigStore: IEndpointConfigStore;
  contextProvider: IContextProvider;
  clock: IClock;
  trustService: TrustService;
  revocationService: RevocationService;
}

export class VerificationService {
  constructor(private readonly deps: VerificationServiceDeps) {}

  async verify(bytes: Uint8Array, options: VerifyOptions = {}): Promise<VerificationResult> {
    const result = this.emptyResult();

    let parsed;
    try {
      parsed = parseSignedObject(bytes);
    } catch (error) {
      result.cryptographic = 'invalid';
      result.message = error instanceof Error ? error.message : 'malformed signed object';
      return this.finalize(result);
    }
    if (parsed.type !== 'sdoc') {
      result.cryptographic = 'invalid';
      result.message = `expected an SDoc, got ${parsed.type}`;
      return this.finalize(result);
    }
    result.sdocId = sdocIdOf(bytes);

    /* ---------- TCert ---------- */
    const keyId = parsed.signerKeyId;
    let tcertNumber: number;
    try {
      tcertNumber = tcertNumberOf(parsed);
    } catch {
      result.tcert = 'invalid';
      result.message = 'SDoc has no valid tcertNumber';
      return this.finalize(result);
    }
    const tcertId = tcertIdOf(keyId, tcertNumber);
    result.tcertId = tcertId;

    const tcertBytes = await this.deps.certificateStore.get(tcertId);
    if (!tcertBytes) {
      result.tcert = 'cannotVerify';
      result.message = 'TCert not found locally';
      return this.finalize(result);
    }
    let tcertParsed;
    try {
      tcertParsed = parseSignedObject(tcertBytes);
    } catch {
      result.tcert = 'invalid';
      result.message = 'stored TCert is malformed';
      return this.finalize(result);
    }
    const tcertProvider = this.deps.cryptoRegistry.get(tcertParsed.algorithm);
    const tcertPub = tcertParsed.data.publicKey as unknown as PublicJwk;
    const selfOk = await verifyParsedSignedObject(tcertParsed, tcertProvider, tcertPub);
    const keyIdOk = tcertProvider.keyId(tcertPub) === tcertParsed.signerKeyId;
    if (!selfOk || !keyIdOk) {
      result.tcert = 'invalid';
      result.message = 'TCert self-signature invalid';
      return this.finalize(result);
    }
    const now = options.currentTime ?? this.deps.clock.now();
    const validity = tcertParsed.data.validity as
      | { validAfter?: number; validBefore?: number; sdocMaxAgeSeconds?: number }
      | undefined;
    if (validity) {
      if (validity.validAfter !== undefined && now < validity.validAfter) {
        result.tcert = 'invalid';
        result.message = 'TCert is not yet valid';
        return this.finalize(result);
      }
      if (validity.validBefore !== undefined && now >= validity.validBefore) {
        result.tcert = 'invalid';
        result.message = 'TCert has expired';
        return this.finalize(result);
      }
      // The TCert can also restrict how long the SDocs it issues stay valid.
      if (validity.sdocMaxAgeSeconds !== undefined) {
        const sdocIssuedAt = parsed.data.issuedAt;
        if (typeof sdocIssuedAt === 'number' && Number.isInteger(sdocIssuedAt) && now - sdocIssuedAt > validity.sdocMaxAgeSeconds) {
          result.schema = 'invalid';
          result.message = 'SDoc exceeds the validity duration set by its TCert';
          return this.finalize(result);
        }
      }
    }
    result.tcert = 'valid';

    /* ---------- Trust ---------- */
    const trust = await this.deps.trustService.resolveTrust(tcertId, tcertParsed);
    result.trust = trust.state;
    if (trust.message) result.message = trust.message;
    if (trust.state === 'invalid') return this.finalize(result);

    /* ---------- Revocation ---------- */
    const issuedAt = parsed.data.issuedAt;
    if (typeof issuedAt !== 'number' || !Number.isInteger(issuedAt)) {
      result.revocation = 'invalid';
      result.message = 'SDoc has no valid issuedAt';
      return this.finalize(result);
    }
    const revocation = await this.deps.revocationService.checkRevocation(tcertId, keyId, issuedAt, result.sdocId!);
    result.revocation = revocation.state;
    if (revocation.message) result.message = revocation.message;
    if (revocation.state === 'invalid') return this.finalize(result);

    /* ---------- Secrets + signature ---------- */
    const rawSchema = tcertParsed.data.schema;
    if (!Array.isArray(rawSchema) || rawSchema.length === 0) {
      result.schema = 'invalid';
      result.message = 'signer TCert has no document schema';
      return this.finalize(result);
    }
    const fields = rawSchema as unknown as FieldSchema[];
    const secretValues: Record<string, string> = {};
    let missingSecret = false;
    for (const field of fields) {
      if (isStrippedBinding(field)) {
        const secret = await this.deps.contextProvider.requestSecret(field);
        if (secret === null) {
          missingSecret = true;
          result.warnings.push(`missing bound value '${field.name}'`);
        } else {
          secretValues[field.name] = secret;
        }
      }
    }
    if (missingSecret) {
      result.cryptographic = 'cannotVerify';
      result.context = 'missing';
      result.message = 'missing required secret input';
      return this.finalize(result);
    }
    const externalAad = Object.keys(secretValues).length > 0 ? cborEncode(secretValues) : new Uint8Array(0);
    const provider = this.deps.cryptoRegistry.get(parsed.algorithm);
    const sigOk = await verifyParsedSignedObject(parsed, provider, tcertPub, externalAad);
    if (!sigOk) {
      result.cryptographic = 'invalid';
      result.message = 'SDoc signature verification failed (tampered data or incorrect secret)';
      return this.finalize(result);
    }
    result.cryptographic = 'valid';

    /* ---------- Schema + fields ---------- */
    const storedValues = parsed.data.fields as unknown[]; // schema-indexed values
    result.schema = 'valid';
    const ctx = this.deps.contextProvider.buildContext();
    for (let i = 0; i < fields.length; i++) {
      const field = fields[i];
      if (!field) continue;
      if (isStrippedBinding(field)) {
        result.fields.push({ name: field.name, label: field.label, state: 'valid', message: 'covered by signature (not stored)' });
        continue;
      }
      const engine = this.deps.fieldRegistry.get(field.type);
      const encoded = storedValues[i];
      if (encoded === undefined || encoded === null) {
        const required = field.inputRules?.required === true;
        if (required) {
          result.schema = 'invalid';
          result.fields.push({ name: field.name, label: field.label, state: 'invalid', message: 'missing required field' });
        } else {
          result.fields.push({ name: field.name, label: field.label, state: 'valid', message: 'absent (optional)' });
        }
        continue;
      }

      // Inline-bound fields: the verifier must re-enter the exact stored value.
      // The signature already verified, so a mismatch is a rules-level failure and
      // the original (stored) value can still be shown.
      if (isBoundField(field) && effectiveBinding(field) === 'inline') {
        const bound = await this.deps.contextProvider.requestSecret(field);
        if (bound === null) {
          result.fields.push({ name: field.name, label: field.label, state: 'cannotVerify', message: 'bound value not provided' });
          result.context = 'missing';
          continue;
        }
        const presentable = String(engine.decode(field, encoded));
        if (!constantTimeEqual(bound.trim(), presentable.trim())) {
          result.schema = 'invalid';
          result.fields.push({ name: field.name, label: field.label, state: 'invalid', message: 'binding value mismatch' });
          continue;
        }
        // Value matches — fall through to the engine's normal validation below.
      }

      // Attachments are compact content-hash strings — the field engine validates
      // the shape. The actual file is NOT downloaded during verification (offline
      // first); the verifier fetches it on demand by hash and checks the hash.
      let fieldResult: FieldResult;
      try {
        fieldResult = await engine.validateField(field, encoded, ctx);
      } catch (error) {
        result.schema = 'invalid';
        result.fields.push({
          name: field.name,
          label: field.label,
          state: 'malformed',
          message: error instanceof Error ? error.message : 'malformed field value',
        });
        continue;
      }
      result.fields.push({ ...fieldResult, label: field.label });
      if (fieldResult.state === 'invalid') result.schema = 'invalid';
      if (fieldResult.state === 'cannotVerify' || fieldResult.state === 'missingContext') result.context = 'missing';
      if (fieldResult.state === 'contextDenied') result.context = 'denied';
    }

    return this.finalize(result);
  }

  private emptyResult(): VerificationResult {
    return {
      overall: 'cannotVerify',
      cryptographic: 'cannotVerify',
      tcert: 'cannotVerify',
      trust: 'cannotVerify',
      revocation: 'cannotVerify',
      schema: 'cannotVerify',
      fields: [],
      context: 'satisfied',
      warnings: [],
    };
  }

  private finalize(result: VerificationResult): VerificationResult {
    const components: ComponentState[] = [
      result.cryptographic,
      result.tcert,
      result.trust,
      result.revocation,
      result.schema,
    ];
    if (components.includes('invalid')) result.overall = 'invalid';
    else if (components.includes('cannotVerify') || result.context === 'missing' || result.context === 'denied') {
      result.overall = 'cannotVerify';
    } else {
      result.overall = 'valid';
    }
    return result;
  }
}

File: qrs-core-python/src/qrs/services/verificationService.py
"""VerificationService: the verification pipeline.

Pipeline:

1. parse the SDoc
2. resolve and verify the TCert (self-signature, id, validity)
3. resolve trust (pinning or CA attestation)
4. evaluate revocation (key, TCert, SDoc block)
5. collect secret inputs, rebuild the COSE external AAD, verify the signature
6. validate the payload against the schema (including contextual fields)
7. produce a structured VerificationResult

The result deliberately distinguishes ``valid``, ``invalid`` and
``cannotVerify`` (e.g. GPS unavailable is NOT the same as "outside the
permitted area").
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ..cbor import cbor_encode
from ..deps import ServiceDeps
from ..envelope import (
    parse_signed_object,
    sdoc_id_of,
    tcert_id_of,
    tcert_number_of,
    verify_parsed_signed_object,
)
from ..errors import QrsValidationError
from ..fields import (
    FieldResult,
    FieldSchema,
    effective_binding,
    is_bound_field,
    is_stripped_binding,
)
from ..id import to_hex

__all__ = ["VerificationService", "VerificationResult", "VerifyOptions", "VerificationServiceDeps"]


@dataclass
class VerificationServiceDeps:
    """Dependencies for the verification pipeline.

    ``trust_service`` and ``revocation_service`` are injected separately because
    verification composes them (and they in turn depend on the same stores).
    """

    crypto_registry: Any
    field_registry: Any
    certificate_store: Any
    endpoint_config_store: Any
    context_provider: Any
    clock: Any
    trust_service: Any
    revocation_service: Any

    @classmethod
    def from_service_deps(cls, deps: ServiceDeps, trust_service: Any, revocation_service: Any) -> "VerificationServiceDeps":
        return cls(
            crypto_registry=deps.crypto_registry,
            field_registry=deps.field_registry,
            certificate_store=deps.certificate_store,
            endpoint_config_store=deps.endpoint_config_store,
            context_provider=deps.context_provider,
            clock=deps.clock,
            trust_service=trust_service,
            revocation_service=revocation_service,
        )


@dataclass
class VerificationResult:
    overall: str = "cannotVerify"
    cryptographic: str = "cannotVerify"
    tcert: str = "cannotVerify"
    trust: str = "cannotVerify"
    revocation: str = "cannotVerify"
    schema: str = "cannotVerify"
    fields: list[FieldResult] = field(default_factory=list)
    context: str = "satisfied"
    warnings: list[str] = field(default_factory=list)
    sdoc_id: str | None = None
    tcert_id: str | None = None
    message: str | None = None


@dataclass
class VerifyOptions:
    current_time: int | None = None


class VerificationService:
    def __init__(self, deps: VerificationServiceDeps) -> None:
        self._deps = deps

    async def verify(self, data: bytes, options: VerifyOptions | None = None) -> VerificationResult:
        options = options or VerifyOptions()
        result = VerificationResult()

        try:
            parsed = parse_signed_object(data)
        except Exception as exc:
            result.cryptographic = "invalid"
            result.message = str(exc) if isinstance(exc, Exception) else "malformed signed object"
            return self._finalize(result)

        if parsed.type != "sdoc":
            result.cryptographic = "invalid"
            result.message = f"expected an SDoc, got {parsed.type}"
            return self._finalize(result)
        result.sdoc_id = sdoc_id_of(data)

        # ---------- TCert ----------
        # The TCert linkage comes from the COSE protected headers: keyId from
        # `kid`, certificate number from the protocol-private `tcertNumber`
        # header. This matches the reference implementation.
        key_id = parsed.signer_key_id
        try:
            tcert_number = tcert_number_of(parsed)
        except Exception:
            result.tcert = "invalid"
            result.message = "SDoc has no valid tcertNumber"
            return self._finalize(result)
        tcert_id = tcert_id_of(key_id, tcert_number)
        result.tcert_id = tcert_id

        tcert_bytes = await self._deps.certificate_store.get(tcert_id)
        if not tcert_bytes:
            result.tcert = "cannotVerify"
            result.message = "TCert not found locally"
            return self._finalize(result)
        try:
            tcert_parsed = parse_signed_object(tcert_bytes)
        except Exception:
            result.tcert = "invalid"
            result.message = "stored TCert is malformed"
            return self._finalize(result)
        tcert_provider = self._deps.crypto_registry.get(tcert_parsed.algorithm)
        tcert_pub = tcert_parsed.data["publicKey"]
        self_ok = verify_parsed_signed_object(tcert_parsed, tcert_provider, tcert_pub)
        key_id_ok = tcert_provider.key_id(tcert_pub) == to_hex(tcert_parsed.data["keyId"])
        if not self_ok or not key_id_ok:
            result.tcert = "invalid"
            result.message = "TCert self-signature invalid"
            return self._finalize(result)

        now = options.current_time if options.current_time is not None else self._deps.clock.now()
        validity = tcert_parsed.data.get("validity") or {}
        if validity:
            if validity.get("validAfter") is not None and now < validity["validAfter"]:
                result.tcert = "invalid"
                result.message = "TCert is not yet valid"
                return self._finalize(result)
            if validity.get("validBefore") is not None and now >= validity["validBefore"]:
                result.tcert = "invalid"
                result.message = "TCert has expired"
                return self._finalize(result)
            if validity.get("sdocMaxAgeSeconds") is not None:
                sdoc_issued_at = parsed.data.get("issuedAt")
                if isinstance(sdoc_issued_at, int) and now - sdoc_issued_at > validity["sdocMaxAgeSeconds"]:
                    result.schema = "invalid"
                    result.message = "SDoc exceeds the validity duration set by its TCert"
                    return self._finalize(result)
        result.tcert = "valid"

        # ---------- Trust ----------
        trust = await self._deps.trust_service.resolve_trust(tcert_id, tcert_parsed)
        result.trust = trust.state
        if trust.message:
            result.message = trust.message
        if trust.state == "invalid":
            return self._finalize(result)

        # ---------- Revocation ----------
        issued_at = parsed.data.get("issuedAt")
        if not isinstance(issued_at, int):
            result.revocation = "invalid"
            result.message = "SDoc has no valid issuedAt"
            return self._finalize(result)
        revocation = await self._deps.revocation_service.check_revocation(
            tcert_id, key_id, issued_at, result.sdoc_id or ""
        )
        result.revocation = revocation.state
        if revocation.message:
            result.message = revocation.message
        if revocation.state == "invalid":
            return self._finalize(result)

        # ---------- Secrets + signature ----------
        raw_schema = tcert_parsed.data.get("schema")
        if not isinstance(raw_schema, list) or len(raw_schema) == 0:
            result.schema = "invalid"
            result.message = "signer TCert has no document schema"
            return self._finalize(result)
        fields = [FieldSchema.from_json(f) for f in raw_schema]
        secret_values: dict[str, str] = {}
        missing_secret = False
        for field in fields:
            if is_stripped_binding(field):
                secret = await self._deps.context_provider.request_secret(field)
                if secret is None:
                    missing_secret = True
                    result.warnings.append(f"missing bound value '{field.name}'")
                else:
                    secret_values[field.name] = secret
        if missing_secret:
            result.cryptographic = "cannotVerify"
            result.context = "missing"
            result.message = "missing required secret input"
            return self._finalize(result)
        external_aad = cbor_encode(secret_values) if secret_values else b""
        provider = self._deps.crypto_registry.get(parsed.algorithm)
        sig_ok = verify_parsed_signed_object(parsed, provider, tcert_pub, external_aad)
        if not sig_ok:
            result.cryptographic = "invalid"
            result.message = "SDoc signature verification failed (tampered data or incorrect secret)"
            return self._finalize(result)
        result.cryptographic = "valid"

        # ---------- Schema + fields ----------
        stored_values = parsed.data.get("fields") or []
        result.schema = "valid"
        ctx = self._deps.context_provider.build_context()
        for i, field in enumerate(fields):
            if is_stripped_binding(field):
                result.fields.append(
                    FieldResult(name=field.name, state="valid", message="covered by signature (not stored)", label=field.label)
                )
                continue
            engine = self._deps.field_registry.get(field.type)
            encoded = stored_values[i] if i < len(stored_values) else None
            if encoded is None:
                required = bool((field.input_rules or {}).get("required"))
                if required:
                    result.schema = "invalid"
                    result.fields.append(
                        FieldResult(name=field.name, state="invalid", message="missing required field", label=field.label)
                    )
                else:
                    result.fields.append(
                        FieldResult(name=field.name, state="valid", message="absent (optional)", label=field.label)
                    )
                continue

            # Inline-bound fields: the verifier must re-enter the exact stored value.
            if is_bound_field(field) and effective_binding(field) == "inline":
                bound = await self._deps.context_provider.request_secret(field)
                if bound is None:
                    result.fields.append(
                        FieldResult(name=field.name, state="cannotVerify", message="bound value not provided", label=field.label)
                    )
                    result.context = "missing"
                    continue
                presentable = str(engine.decode(field, encoded))
                if bound.strip() != presentable.strip():
                    result.schema = "invalid"
                    result.fields.append(
                        FieldResult(name=field.name, state="invalid", message="binding value mismatch", label=field.label)
                    )
                    continue

            try:
                field_result = await engine.validate_field(field, encoded, ctx)
            except Exception as exc:
                result.schema = "invalid"
                result.fields.append(
                    FieldResult(name=field.name, state="malformed", message=str(exc), label=field.label)
                )
                continue
            result.fields.append(field_result)
            if field_result.state == "invalid":
                result.schema = "invalid"
            if field_result.state in ("cannotVerify", "missingContext"):
                result.context = "missing"
            if field_result.state == "contextDenied":
                result.context = "denied"

        return self._finalize(result)

    def _finalize(self, result: VerificationResult) -> VerificationResult:
        components = [result.cryptographic, result.tcert, result.trust, result.revocation, result.schema]
        if "invalid" in components:
            result.overall = "invalid"
        elif "cannotVerify" in components or result.context in ("missing", "denied"):
            result.overall = "cannotVerify"
        else:
            result.overall = "valid"
        return result